src/main/scala/edu/ie3/osmogrid/lv/region_coordinator/BoundaryFactory.scala

```diff
@@ -65,120 +66,191 @@
-          s"Empty boundary relation ${relation.id}."
-        )
-      case nodes => nodes
-    })
-      .map { node =>
-        // Turn node ids into nodes
+
+        (Seq(resultNodes))
+      }
+      .flatten
+
+    val listOfSeq: Map[Int, Seq[AreaKey]] = splitByKey(allNodesOfPolygonsSeq)
+    val polygonList: ListBuffer[Polygon] = ListBuffer.empty[Polygon]
+
+    listOfSeq.values.foreach { seq =>
+      val latLonCoordinates: Seq[Coordinate] = seq.flatMap { node =>
         enhancedRelation
           .node(node)
-          .getOrElse(
-            throw new RuntimeException(
-              s"Node $node not found in enhanced relation $enhancedRelation"
-            )
-          )
+          .map(n => GeoUtils.buildCoordinate(n.latitude, n.longitude))
       }
-      .map { node =>
-        // Turn nodes into coordinates
-        GeoUtils.buildCoordinate(node.latitude, node.longitude)
+
+      // Ensure the LinearRing is closed
+      val closedCoordinates =
+        if (
+          latLonCoordinates.nonEmpty &&
+          latLonCoordinates.headOption == latLonCoordinates.lastOption
+        )
+          latLonCoordinates
+        else
+          throw new RuntimeException(
+            s"First node should be last in boundary relation ${relation.id}."
+          )
+
+      val polygon: Polygon = GeoUtils.buildPolygon(closedCoordinates.toArray)
+      polygonList += polygon
+    }
+
+    polygonList.toList
+
+  }
+
+  /** Split node sequences of polygons into parts. If the sequence can't be
+    * split because it represents only one polygon, the sequence is returned
+    * with the first node added at the end to close the polygon.
+    *
+    * @param seq
+    *   Sequence to split
+    * @tparam A
+    *   type of sequence
+    * @return
+    *   Indexed map with the split sequences
+    */
+
+  private def splitByKey[A](seq: Seq[A]): Map[Int, Seq[A]] = {
+    val result = seq
+      .foldLeft((Map.empty[Int, Seq[A]], Map.empty[A, Int], 0)) {
+        case ((acc, indexes, splitIdx), elem) =>
+          indexes.get(elem) match {
+            case Some(prevIdx) =>
+              // Element has been seen before
+              val newSeq = seq.slice(prevIdx, splitIdx + 1)
+              (
+                acc + (acc.size + 1 -> newSeq),
+                indexes + (elem -> splitIdx),
+                splitIdx + 1,
+              )
+            case None =>
+              // Element has not been seen before
+              (
+                acc,
+                indexes + (elem -> splitIdx),
+                splitIdx + 1,
+              )
+          }
       }
-      .toArray
+      ._1
 
-    GeoUtils.buildPolygon(coordinates)
+    // If the result map is empty, return the whole sequence with the first element appended at the end
+    // (to close the polygon)
+    if (result.isEmpty) {
+      seq.headOption match {
+        case Some(firstElem) => Map(1 -> (seq :+ firstElem))
+        case None =>
+          Map(1 -> seq)
```

Comment:
You could return an empty map here, since seq is empty in this case.

Reply:
Good catch

Reply:
There you go

Reply:
Thanks.